File: doc/src/codes/Users_Manual/advanced_examples/custom_gain_0.cpp
#include "autd3.hpp"

class FocalPoint final : public autd3::Gain {
 public:
  explicit FocalPoint(autd3::Vector3 point) : _point(std::move(point)) {}

  std::unordered_map<size_t, std::vector<autd3::Drive>> calc(
      const autd3::Geometry& geometry) const override {
    return autd3::Gain::transform(
        geometry, [&](const auto& dev, const auto& tr) {
          const auto phase = (tr.position() - _point).norm() *
                             tr.wavelength(dev.sound_speed());
          return autd3::Drive{autd3::Phase::from_rad(phase),
                              autd3::EmitIntensity::maximum()};
        });
  }

 private:
  autd3::Vector3 _point;
};

File: doc/src/codes/Users_Manual/advanced_examples/custom_gain_0.rs
# extern crate autd3;
# extern crate autd3_driver;
use std::collections::HashMap;
use autd3::{
    derive::Gain,
    prelude::*,
};
use autd3_driver::derive::prelude::*;

#[derive(Gain)]
pub struct FocalPoint {
    position: Vector3,
}

impl FocalPoint {
    pub fn new(position: Vector3) -> Self {
        Self {position}
    }
}

impl Gain for FocalPoint {
    fn calc(&self, geometry: &Geometry, filter: GainFilter) -> Result<HashMap<usize, Vec<Drive>>, AUTDInternalError> {
        Ok(Self::transform(geometry, filter, |dev, tr| Drive {
            phase: Phase::from_rad((tr.position() - self.position).norm() * tr.wavelength(dev.sound_speed)),
            intensity: EmitIntensity::MAX,
        }))
    }
}
# fn main() { 
# }
#